feat: Implement Locator.nth() (#7137)
vitest-dev/vitest: docs/guide/browser/locators.md

- [testing-library's `ByTestId`](https://testing-library.com/docs/queries/bytestid/)

## nth

```ts
function nth(index: number): Locator
```

> This method returns a new locator that matches only a specific index within a multi-element query result. This is similar to calling `elements()[n]` but easier to use with asynchronous retry.

Reviewer: ```suggestion
This method returns a new locator that matches only a specific index within a multi-element query result. Unlike `elements()[n]`, the `nth` locator will be retries until the element is present.
```

```html
<div aria-label="one"><input/><input/><input/></div>
<div aria-label="two"><input/></div>
```

```tsx
page.getByRole('textbox').nth(0) // ✅
page.getByRole('textbox').nth(4) // ❌
```

::: tip
Before resorting to `nth`, you may find it useful to use chained locators to narrow down your search.
Sometimes there is no better way to distinguish than by element position; although this can lead to flake, it's better than nothing.
:::

```tsx

> page.getByLabel('two').getByRole('input') // ✅ better alternative to page.nth(3)

Reviewer: `page.nth()` is not a thing (and shouldn't be). I do wonder if we should expose `page.body` as `Locator`, but this is out of scope of this PR

Author: D'oh! Good catch. Fixing. (I assume you like the idea of annotating the ✅/❌ to explain the nuance.)

Re `page.body`, sounds like an excellent discussion to have on the Discord, or in an issue. (Along with `and` / `or` and the idea of an extension API...)

page.getByLabel('one').getByRole('input') // ❌ too ambiguous
page.getByLabel('one').getByRole('input').nth(1) // ✅ pragmatic compromise
```

## first

```ts
function first(): Locator
```

This method returns a new locator that matches only the first index of a multi-element query result.
It is sugar for `nth(0)`.

```html
<input/> <input/> <input/>
```

```tsx

> page.getByRole('textbox').first() // ✅

Reviewer 2: I am trying to use this in my project
 await expect.element(screen.getByText("Org 1").first()).toBeVisible();
 
 test succeeds but there is a type error
 ```text
src/pages/Admin/PageAdminIssueReports.spec.ts:178:56 - error TS2339: Property 'first' does not exist on type 'Locator'.

178         await expect.element(screen.getByText("Org 1").first()).toBeVisible();

 ```
 Do I need to include some special types for this?

Reviewer: Good point. All public methods need to be defined also here: https://github.com/vitest-dev/vitest/blob/main/packages/browser/context.d.ts

Author: Whoops; sorry about that. Addressed in #7176

- [See `locator.elements()`](#elements)

> ### nth, first, last

Reviewer: these need to be separate and have examples

they should also come after all locator-returning methods

Author: I'll separate them and add examples.

They currently come after `all` which is a locator-returning method of `Locator`. I figured that the `getByXyz` were in their own group at the top because, although they are methods, they are also root locators of `Page` whereas `all`, `nth`, etc are only methods.

Was your goal for me to move `nth`, `first` and `last` to the top, below the `getXyz`, but leave `all` where it is? Or should they remain below `all`? (Or something else? Just LMK.)

Reviewer: just move yours bellow `get`. `all` returns an array, not a locator
